### PR TITLE
Topic recognition - s4647936

### PR DESCRIPTION
**Pull request by tranvicky**
This pull request introduces an Alzheimer's Disease classification solution using Siamese networks on the ADNI dataset. The Siamese Network, with a Simple Classifier, allows for the differentiation between Alzheimer's Disease (AD) and Normal Control (NC) brain images. The primary objective is to use triplet loss and the power of Siamese networks to achieve this classification.

**Key Highlights:**
- Implementation of a Siamese network paired with a simple classifier for AD vs. NC classification.
- The use of the TripletDataset for generating triplets of anchor, positive, and negative images for training.
- Data augmentation techniques to enhance model robustness during training.
- Visualisation of embeddings post-training using t-SNE.
- In-depth analysis using confusion matrices to evaluate the performance of the classifier on test embeddings.

**Files Added:**

- `train.py`: Script containing the Siamese network training logic, visualisation of results, and classifier training.
- `modules.py`: Contains the architecture for the Feature Extractor, Siamese Network, Triplet Loss, and the Simple Classifier.
- `dataset.py`: Provides the TripletDataset class for generating triplets and the logic for a patient-wise dataset split.
- `predict.py`: Script to generate embeddings using the trained Siamese Network, visualise them, and evaluate the classifier's performance.
- Various plots and figures showcasing the t-SNE visualisations, training vs. validation loss, and confusion matrices.

**Setup and Requirements:**
- Requires PyTorch 1.10.1+cu113 or above.
- Dependencies include torchvision, Pillow, matplotlib, seaborn, numpy, and scikit-learn.
- Detailed setup, installation, and usage instructions can be found in the README.

Thanks,
Vicky (s4647936)

### CODE REVIEW

**Comment by nathasha-naranpanawa** [comment]
**This is an initial inspection, no action is required at this point**

Difficulty: Hard

Readme: Very good
- Clear flow of information
- Algortihm and usage explained well
- Relevant plots for loss and 2D manifolds provided

Commit messages: very good, detailed 

Code: 
- uses a triplet network
- good design
- sufficiently commented

Functionality/Performance: 
- the triplet network seem to have converged but the manifold doesn't look great
- classifier seem to perform poorly (no accuracy plot/average accuracy for the classfier shown other than the confusion matrix)
- future work have been suggested but no attempts have been made to improve either sub networks

General comments:
- **PR is in the main branch**
- solves the problem apporpriately although performance is poor
- one major reason the triplet isn't doing a good job of separating the two classes could be because your feature extracting backbone network isn't strong enough. Could have tried something like a ResNet or VGG to see if it would improve.

**Comment by shakes76** [comment]
# Marking
## Good Practice (Design/Commenting, TF/Torch Usage)
Adequate design and implementation
Good spacing and comments
Header blocks missing -1
## Recognition Problem 
Solves problem 
Driver Script present
File structure present
Shows Usage & Demo & Visualisation & Data usage
Module present
Commenting
No Data leakage
Difficulty: Hard
## Commit Log
Meaningful commit messages
Progressive commits used
## Documentation
ReadMe acceptable/good
Model/technical explanation
Good Description and Comments 
Markdown used and PDF submitted
## Pull Request
Successful Pull Request (Working Algorithm Delivered on Time in InCorrect Branch) -2
Feedback required, please change the branch to the correct one, ensure repo READMEs are restored. -2
Request Description good

**Comment by shakes76** [comment]
Feedback marks possible +2 if the requested changes are made (see above).

**Comment by wangzhaomxy** [comment]
No feedback attempt and no feedback marks granted.
